dev-lang/python/patches/python3.10-3.10.14.patchset

```diff
@@ -750,7 +532,7 @@ index ec3cece..6a29d2a 100755
 2.45.2
 
 
-From 390b9112392e602714714c8bb0275beee85914fe Mon Sep 17 00:00:00 2001
+From a5f44dabeeae79d50ad890d398e3a531917bb3c6 Mon Sep 17 00:00:00 2001
 From: Jerome Duval <jerome.duval@gmail.com>
 Date: Fri, 4 Oct 2019 22:02:35 +0200
 Subject: since 3.8, don't reinit locks on fork.
```

Comment:
Completely unrelated to this change, but:

I haven't looked at the corresponding code exactly, but this boosts my confidence that my "scary" patch for perl (https://github.com/haikuports/haikuports/pull/10626/files#diff-0870d8eb3e20e6a6321e6e8e4c72df07fcf42e600ddd6bd53491fbcd1ceb904dR536-R568) is actually correct on Haiku (judging from that commit message).

Reply:
Mh, this doesn't sound right to me. Why doesn't it work just the same as other POSIX OSes? We acquire all the heap locks in libroot and unlock them after fork.

Reply:
The EPERM would probably be because the child has a new thread ID and is trying to unlock a mutex that is locked by the parent thread. I think it should reinit the locks instead, but it looks like that's what this code does...?

The perl code is more interesting. Does it really just try to unlock the locks in the child? I'm not sure how that would work on Linux either.

Reply:
I'm not sure if these two problems (python and perl) are related at all.

If I decoded the error number correctly, I'm pretty sure I got EPERM when building perl.

Reply:
From what I am seeing, perl on `musl` would have the same problem -- unless it uses `PTHREAD_MUTEX_NORMAL`, it appears? So maybe we need to adjust Haiku's behavior for that.

Reply:
miniperl apparently calls the unlock function in both the parent and the child after fork. See https://github.com/Perl/perl5/blob/d8935409c94992d3e665cdaa1c5411eb55a0e9e2/miniperlmain.c#L87-L100

Reply:
It looks like Perl is using `pthread_mutexattr_default`, which means we get `PTHREAD_MUTEX_DEFAULT`. The specification indicates that implementations are allowed to map this to one of the other types, and Haiku basically has it behave like `PTHREAD_MUTEX_ERRORCHECK`, so we get this behavior.

Reply:
Ultimately your patch probably isn't safe though, the right thing to do here is to re-initialize the locks in the child. This probably qualifies as a bug in Perl because it is relying on behavior that the POSIX specification explicitly states is undefined (see https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_mutex_unlock.html). I would report it upstream, honestly.

Reply:
This code in Python looks correct to me however. @korli, any recollection of why you added this patch?

Reply:
https://github.com/haikuports/haikuports/commit/2993b575fc087a4c91c087fa4f00b299473a1eb4

So, we should retest with `muon test` I suppose and see what happens on a nightly build.

Reply:
You mean "meson test", right?

Reply:
Yes. May be possible that the hangs were related to bugs in Haiku's locks that have since been fixed.

Reply:
> Ultimately your patch probably isn't safe though, the right thing to do here is to re-initialize the locks in the child. This probably qualifies as a bug in Perl because it is relying on behavior that the POSIX specification explicitly states is undefined (see https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html). I would report it upstream, honestly.

It looks like there is already a ~10 year old issue about this in perl: https://github.com/Perl/perl5/issues/13237 with an old Debian patch for that. Although it seems that Debian (kFreeBSD) doesn't seem to be affected by it any more.